File: ios/DesMoinesInsider/Models/Restaurant.swift
import Foundation
import CoreLocation

struct Restaurant: Identifiable, Codable, Hashable {
    let id: String
    let name: String
    var cuisine: String?
    var location: String?
    var city: String?
    var rating: Double?
    var priceRange: String?
    var description: String?
    var phone: String?
    var website: String?
    var imageUrl: String?
    var isFeatured: Bool?
    var latitude: Double?
    var longitude: Double?
    var popularityScore: Double?
    var status: String?
    var slug: String?
    var aiWriteup: String?
    var businessHours: BusinessHours?
    var dietaryOptions: [String]?
    var createdAt: String?
    var updatedAt: String?
    /// First-party sponsored-listing flag (IOS-ADS-011). Set by the backend
    /// while a paid sponsorship is active; `sponsored_until` is informational.
    var isSponsored: Bool?
    var sponsoredUntil: String?

    enum CodingKeys: String, CodingKey {
        case id, name, cuisine, location, city, rating, description, phone, website, status, slug
        case priceRange = "price_range"
        case imageUrl = "image_url"
        case isFeatured = "is_featured"
        case latitude, longitude
        case popularityScore = "popularity_score"
        case aiWriteup = "ai_writeup"
        case businessHours = "business_hours"
        case dietaryOptions = "dietary_options"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case isSponsored = "is_sponsored"
        case sponsoredUntil = "sponsored_until"
    }

    /// Whether this listing currently carries an active paid sponsorship
    /// (IOS-ADS-011). Mirrors the web, which keys off the `is_sponsored` flag.
    var isActivelySponsored: Bool { isSponsored == true }

    // MARK: - Open/Closed Status

    /// Determines if the restaurant is currently open based on business_hours.
    /// Returns nil if hours data is unavailable.
    /// Shared calendar so the "Open Now" filter doesn't allocate Calendar.current
    /// for every restaurant on every toggle/page — the dominant per-call cost on
    /// a 100+ list (IOS-AUDIT-PERF-009). Captures the user's timezone once.
    private static let sharedCalendar = Calendar.current

    func isOpenNow(at date: Date = .now) -> Bool? {
        guard let hours = businessHours else { return nil }

        // One dateComponents call instead of three component(_:from:) calls.
        let parts = Self.sharedCalendar.dateComponents([.weekday, .hour, .minute], from: date)
        guard let weekday = parts.weekday, let hour = parts.hour, let minute = parts.minute else {
            return nil
        }
        let dayName = Self.dayNames[weekday - 1] // 1=Sun, 2=Mon...

        guard let dayHours = hours.hours(for: dayName) else { return nil }

        // The parsed open/close range is memoized by the hours string so the
        // string-splitting + AM/PM parsing doesn't re-run for every restaurant on
        // every "Open Now" evaluation (IOS-AUDIT-PERF-021).
        switch Self.dayRange(for: dayHours) {
        case .closed:
            return false
        case .unknown:
            return nil
        case .open(let open, let close):
            let currentMinutes = hour * 60 + minute
            // Handle overnight hours (close < open means past midnight)
            if close < open {
                return currentMinutes >= open || currentMinutes < close
            }
            return currentMinutes >= open && currentMinutes < close
        }
    }

    /// Parsed open/close minutes for a day's hours string.
    private enum DayRange { case open(Int, Int), closed, unknown }

    private static let rangeCacheLock = NSLock()
    private static var rangeCache: [String: DayRange] = [:]

    /// Memoized parse of a "11:00 AM - 10:00 PM"-style string into minutes.
    /// Thread-safe: the Open Now filter runs on a detached task.
    private static func dayRange(for dayHours: String) -> DayRange {
        rangeCacheLock.lock()
        if let cached = rangeCache[dayHours] {
            rangeCacheLock.unlock()
            return cached
        }
        rangeCacheLock.unlock()

        let computed: DayRange
        let trimmed = dayHours.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty || trimmed.lowercased() == "closed" {
            computed = .closed
        } else {
            let components = dayHours.components(separatedBy: " - ")
            if components.count == 2,
               let open = parseTime(components[0].trimmingCharacters(in: .whitespaces)),
               let close = parseTime(components[1].trimmingCharacters(in: .whitespaces)) {
                computed = .open(open, close)
            } else {
                computed = .unknown
            }
        }

        rangeCacheLock.lock()
        rangeCache[dayHours] = computed
        rangeCacheLock.unlock()
        return computed
    }

    var openStatusText: String {
        switch isOpenNow() {
        case true: return "Open"
        case false: return "Closed"
        case nil: return "Hours unknown"
        }
    }

    private static let dayNames = ["sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday"]

    /// Parse time string like "11:00", "11:00 AM", "9:30 PM" to minutes since midnight.
    private static func parseTime(_ string: String) -> Int? {
        let trimmed = string.trimmingCharacters(in: .whitespaces)

        // Try "HH:mm" (24-hour)
        let parts = trimmed.components(separatedBy: ":")
        guard parts.count >= 2, let hour = Int(parts[0]) else { return nil }

        let minutePart = parts[1].trimmingCharacters(in: .letters).trimmingCharacters(in: .whitespaces)
        guard let minute = Int(minutePart) else { return nil }

        var h = hour
        let upper = trimmed.uppercased()
        if upper.hasSuffix("PM") && h != 12 { h += 12 }
        if upper.hasSuffix("AM") && h == 12 { h = 0 }

        return h * 60 + minute
    }

    // MARK: - Computed Properties

    var coordinate: CLLocationCoordinate2D? {
        // A missing coordinate is null in the DB (Double? == nil); a literal 0.0
        // is a valid location and must not be masked (IOS-AUDIT-BUG-016).
        guard let lat = latitude, let lng = longitude else { return nil }
        return CLLocationCoordinate2D(latitude: lat, longitude: lng)
    }

    var displayDescription: String {
        aiWriteup ?? description ?? ""
    }

    var displayLocation: String {
        [location, city].compactMap { $0 }.filter { !$0.isEmpty }.joined(separator: ", ")
    }

    var ratingText: String {
        guard let rating else { return "No rating" }
        return String(format: "%.1f", rating)
    }

    var priceLevel: Int {
        priceRange?.filter({ $0 == "$" }).count ?? 0
    }

    var callURL: URL? {
        guard let phone, !phone.isEmpty else { return nil }
        let cleaned = phone.replacingOccurrences(of: "[^0-9+]", with: "", options: .regularExpression)
        return URL(string: "tel://\(cleaned)")
    }

    /// Safe http/https website URL only (IOS-AUDIT-SEC-002) — an unsafe scheme
    /// in the content row yields nil so no button renders.
    var websiteURL: URL? {
        website.flatMap { $0.safeWebURL }
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: Restaurant, rhs: Restaurant) -> Bool {
        lhs.id == rhs.id
    }

    // MARK: - Accessibility

    /// Full VoiceOver label for a compact card button (name, cuisine, price range, rating).
    var compactCardAccessibilityLabel: String {
        var parts: [String] = [name]
        if let cuisine { parts.append(cuisine) }
        if let priceRange, !priceRange.isEmpty { parts.append(priceRange) }
        if rating != nil { parts.append("Rated \(ratingText)") }
        return parts.joined(separator: ". ")
    }
}

// MARK: - Business Hours

/// Flexible decoder for the business_hours JSONB column.
/// Supports both `{ "monday": "11:00 - 22:00", ... }` and
/// `{ "hours": { "monday": "..." } }` shapes.
struct BusinessHours: Codable, Hashable {
    private let store: [String: String]

    init(from decoder: Decoder) throws {
        // Try flat { "monday": "...", "tuesday": "..." }
        if let flat = try? decoder.singleValueContainer().decode([String: String].self) {
            store = flat.reduce(into: [:]) { $0[$1.key.lowercased()] = $1.value }
            return
        }
        // Try nested { "hours": { "monday": "..." } }
        if let nested = try? decoder.singleValueContainer().decode([String: [String: String]].self),
           let inner = nested["hours"] {
            store = inner.reduce(into: [:]) { $0[$1.key.lowercased()] = $1.value }
            return
        }
        store = [:]
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(store)
    }

    func hours(for day: String) -> String? {
        store[day.lowercased()]
    }
}

// MARK: - Preview Helpers

extension Restaurant {
    static let preview = Restaurant(
        id: "preview-1",
        name: "Zombie Burger + Drink Lab",
        cuisine: "American",
        location: "300 E Grand Ave",
        city: "Des Moines",
        rating: 4.5,
        priceRange: "$$",
        description: "Creative burgers with horror-themed names and craft cocktails in a fun, quirky atmosphere.",
        phone: "(515) 555-0123",
        website: "https://zombieburger.com",
        isFeatured: true,
        latitude: 41.5910,
        longitude: -93.6088
    )
}

File: android/app/src/main/java/com/desmoines/aipulse/data/model/Restaurant.kt
package com.desmoines.aipulse.data.model

import android.net.Uri
import androidx.compose.runtime.Immutable
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.time.LocalDate
import java.time.LocalTime
import java.util.Locale

/**
 * Restaurant data model matching the Supabase `restaurants` table.
 * Mirrors iOS Restaurant.swift.
 */
@Immutable
@Serializable
data class Restaurant(
    val id: String,
    val name: String,
    val cuisine: String? = null,
    val location: String? = null,
    val city: String? = null,
    val rating: Double? = null,
    @SerialName("price_range") val priceRange: String? = null,
    val description: String? = null,
    val phone: String? = null,
    val website: String? = null,
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("is_featured") val isFeatured: Boolean? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    @SerialName("popularity_score") val popularityScore: Double? = null,
    val status: String? = null,
    val slug: String? = null,
    @SerialName("ai_writeup") val aiWriteup: String? = null,
    @SerialName("business_hours") val businessHours: JsonElement? = null,
    @SerialName("dietary_options") val dietaryOptions: List<String>? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
) {

    // region Open/Closed Status

    companion object {
        private val DAY_NAMES = listOf("sunday", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday")

        /**
         * Parse time string like "11:00", "11:00 AM", "9:30 PM" to minutes since midnight.
         * Mirrors iOS Restaurant.parseTime(_:)
         */
        private fun parseTime(string: String): Int? {
            val trimmed = string.trim()
            val parts = trimmed.split(":")
            if (parts.size < 2) return null
            val hour = parts[0].trim().toIntOrNull() ?: return null
            val minuteStr = parts[1].replace(Regex("[A-Za-z]"), "").trim()
            val minute = minuteStr.toIntOrNull() ?: return null

            var h = hour
            val upper = trimmed.uppercase(Locale.US)
            if (upper.endsWith("PM") && h != 12) h += 12
            if (upper.endsWith("AM") && h == 12) h = 0

            return h * 60 + minute
        }

        /**
         * Get hours string for a given day from the business_hours JSON.
         * Supports both flat {"monday": "11:00 - 22:00"} and nested {"hours": {"monday": "..."}} shapes.
         */
        private fun getHoursForDay(businessHours: JsonElement?, day: String): String? {
            if (businessHours == null || businessHours !is JsonObject) return null
            val obj = businessHours.jsonObject

            // Try flat format
            obj[day]?.let {
                if (it is JsonPrimitive) return it.jsonPrimitive.content
            }

            // Try nested format { "hours": { "monday": "..." } }
            obj["hours"]?.let { hoursEl ->
                if (hoursEl is JsonObject) {
                    hoursEl.jsonObject[day]?.let {
                        if (it is JsonPrimitive) return it.jsonPrimitive.content
                    }
                }
            }

            return null
        }

        val preview = Restaurant(
            id = "preview-1",
            name = "Zombie Burger + Drink Lab",
            cuisine = "American",
            location = "300 E Grand Ave",
            city = "Des Moines",
            rating = 4.5,
            priceRange = "$$",
            description = "Creative burgers with horror-themed names and craft cocktails in a fun, quirky atmosphere.",
            phone = "(515) 555-0123",
            website = "https://zombieburger.com",
            isFeatured = true,
            latitude = 41.5910,
            longitude = -93.6088,
        )
    }

    /**
     * Determines if the restaurant is currently open based on business_hours.
     * Returns null if hours data is unavailable.
     * Mirrors iOS Restaurant.isOpenNow(at:)
     */
    fun isOpenNow(now: java.time.LocalDateTime = java.time.LocalDateTime.now()): Boolean? {
        if (businessHours == null) return null

        val dayOfWeek = now.toLocalDate().dayOfWeek // MONDAY=1 .. SUNDAY=7
        val dayIndex = when (dayOfWeek) {
            java.time.DayOfWeek.SUNDAY -> 0
            java.time.DayOfWeek.MONDAY -> 1
            java.time.DayOfWeek.TUESDAY -> 2
            java.time.DayOfWeek.WEDNESDAY -> 3
            java.time.DayOfWeek.THURSDAY -> 4
            java.time.DayOfWeek.FRIDAY -> 5
            java.time.DayOfWeek.SATURDAY -> 6
            else -> return null
        }
        val dayName = DAY_NAMES[dayIndex]

        val dayHours = getHoursForDay(businessHours, dayName) ?: return null
        if (dayHours.isBlank() || dayHours.lowercase() == "closed") return false

        val components = dayHours.split(" - ")
        if (components.size != 2) return null

        val open = parseTime(components[0]) ?: return null
        val close = parseTime(components[1]) ?: return null

        val currentMinutes = now.hour * 60 + now.minute

        // Handle overnight hours (close < open means past midnight)
        return if (close < open) {
            currentMinutes >= open || currentMinutes < close
        } else {
            currentMinutes >= open && currentMinutes < close
        }
    }

    val openStatusText: String
        get() = when (isOpenNow()) {
            true -> "Open"
            false -> "Closed"
            null -> "Hours unknown"
        }

    // endregion

    // region Computed Properties

    val coordinate: LatLng?
        get() {
            val lat = latitude ?: return null
            val lng = longitude ?: return null
            if (lat == 0.0 && lng == 0.0) return null
            return LatLng(lat, lng)
        }

    val displayDescription: String
        get() = aiWriteup ?: description ?: ""

    val displayLocation: String
        get() = listOfNotNull(location, city).filter { it.isNotEmpty() }.joinToString(", ")

    val ratingText: String
        get() = rating?.let { String.format(Locale.US, "%.1f", it) } ?: "No rating"

    /** Count of '$' characters in priceRange. */
    val priceLevel: Int
        get() = priceRange?.count { it == '$' } ?: 0

    val callUri: Uri?
        get() {
            if (phone.isNullOrEmpty()) return null
            val cleaned = phone.replace(Regex("[^0-9+]"), "")
            return Uri.parse("tel:$cleaned")
        }

    val websiteUri: Uri?
        get() {
            if (website.isNullOrEmpty()) return null
            val url = if (website.startsWith("http")) website else "https://$website"
            return Uri.parse(url)
        }

    /** Full accessibility label for a compact card (name, cuisine, price range, rating). */
    val compactCardAccessibilityLabel: String
        get() {
            val parts = mutableListOf(name)
            cuisine?.let { parts.add(it) }
            if (!priceRange.isNullOrEmpty()) parts.add(priceRange)
            if (rating != null) parts.add("Rated $ratingText")
            return parts.joinToString(". ")
        }

    // endregion
}
